src/java/voldemort/store/rocksdb/RocksDbStorageEngine.java

```diff
@@ -0,0 +1,490 @@
+package voldemort.store.rocksdb;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import org.apache.commons.codec.binary.Hex;
+import org.apache.log4j.Logger;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+import voldemort.VoldemortException;
+import voldemort.store.AbstractStorageEngine;
+import voldemort.store.PersistenceFailureException;
+import voldemort.store.StoreBinaryFormat;
+import voldemort.store.StoreUtils;
+import voldemort.utils.ByteArray;
+import voldemort.utils.ByteUtils;
+import voldemort.utils.ClosableIterator;
+import voldemort.utils.Pair;
+import voldemort.utils.StripedLock;
+import voldemort.versioning.ObsoleteVersionException;
+import voldemort.versioning.Occurred;
+import voldemort.versioning.Version;
+import voldemort.versioning.Versioned;
+
+/**
+ * A StorageEngine that uses RocksDB for persistence
+ * 
+ * 
+ */
+public class RocksDbStorageEngine extends AbstractStorageEngine<ByteArray, byte[], byte[]> {
+
+    private static final Logger logger = Logger.getLogger(RocksDbStorageEngine.class);
+    private RocksDB rocksDB;
+    private final StripedLock locks;
+    private static final Hex hexCodec = new Hex();
+
+    // TODO Need to add stats and loggers later
+
+    public RocksDbStorageEngine(String name, RocksDB rdbInstance, int lockStripes) {
+        super(name);
+        this.rocksDB = rdbInstance;
+        this.locks = new StripedLock(lockStripes);
+    }
+
+    public RocksDB getRocksDB() {
+        return rocksDB;
+    }
+
+    @Override
+    public ClosableIterator<Pair<ByteArray, Versioned<byte[]>>> entries() {
+        return new RocksdbEntriesIterator(this.getRocksDB().newIterator());
+    }
+
+    @Override
+    public ClosableIterator<ByteArray> keys() {
+        return new RocksdbKeysIterator(this.getRocksDB().newIterator());
+    }
+
+    @Override
+    public ClosableIterator<Pair<ByteArray, Versioned<byte[]>>> entries(int partitionId) {
+        throw new UnsupportedOperationException("Partition based entries scan not supported for this storage type");
+    }
+
+    @Override
+    public ClosableIterator<ByteArray> keys(int partitionId) {
+        throw new UnsupportedOperationException("Partition based keys scan not supported for this storage type");
+    }
+
+    @Override
+    public void truncate() {
+        throw new UnsupportedOperationException("truncate not suppported for this storage type");
+    }
+
+    @Override
+    public List<Versioned<byte[]>> get(ByteArray key, byte[] transforms)
+            throws PersistenceFailureException {
+        // TODO read locks ?
+        StoreUtils.assertValidKey(key);
+        long startTimeNs = -1;
+
+        if(logger.isTraceEnabled())
+            startTimeNs = System.nanoTime();
+
+        List<Versioned<byte[]>> value = null;
```

Comment:
can you add a mode where get() also take a lock based on a config value? (bdb has an internal config for this mode).. Check BdbStorageEngine